geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseDataStore.scala

```diff
@@ -0,0 +1,116 @@
+* accompanies this distribution and is available at
+* http://www.opensource.org/licenses/apache2.0.php.
+*************************************************************************/
+package org.locationtech.geomesa.hbase.data
+
+import java.io.Serializable
+import java.util
+
+import org.apache.hadoop.hbase.client._
+import org.apache.hadoop.hbase.util.Bytes
+import org.apache.hadoop.hbase.{HBaseConfiguration, HColumnDescriptor, HTableDescriptor, TableName}
+import org.geotools.data.DataAccessFactory.Param
+import org.geotools.data.store.{ContentDataStore, ContentEntry, ContentFeatureSource}
+import org.geotools.data.{AbstractDataStoreFactory, DataStore, Transaction}
+import org.geotools.feature.NameImpl
+import org.locationtech.geomesa.utils.geotools.SimpleFeatureTypes
+import org.opengis.feature.`type`.Name
+import org.opengis.feature.simple.SimpleFeatureType
+
+import scala.collection.JavaConversions._
+
+class HBaseDataStore(conn: Connection,
+                     catalog: String) extends ContentDataStore {
+
+  import HBaseDataStore._
+
+  private val CATALOG_TABLE = TableName.valueOf(catalog)
+  private val catalogTable = getOrCreateCatalogTable(conn, CATALOG_TABLE)
+  private val schemaCQ = Bytes.toBytes("schema")
+
+  override def createSchema(featureType: SimpleFeatureType): Unit = {
+    val name = featureType.getTypeName
+
+    // create the z3 index
+    val desc = new HTableDescriptor(TableName.valueOf(getZ3TableName(featureType)))
+    Seq(DATA_FAMILY_NAME).foreach { f => desc.addFamily(new HColumnDescriptor(f)) }
+    conn.getAdmin.createTable(desc)
+
+    // write the meta-data
+    val row = Bytes.toBytes(name)
+    val encodedSFT = Bytes.toBytes(SimpleFeatureTypes.encodeType(featureType))
+
+    val schema = new Put(row).addColumn(META_FAMILY_NAME, schemaCQ, encodedSFT)
+    catalogTable.put(schema)
+  }
+
+  def getZ3TableName(sft: SimpleFeatureType) = s"${sft.getTypeName}_z3"
+  def getZ3Table(sft: SimpleFeatureType) = conn.getTable(TableName.valueOf(getZ3TableName(sft)))
+
+  override def createFeatureSource(entry: ContentEntry): ContentFeatureSource = {
+    val sft =
+      Option(entry.getState(Transaction.AUTO_COMMIT).getFeatureType).getOrElse { getSchema(entry) }
+    new HBaseFeatureSource(entry, null, sft)
+  }
+
+  def getSchema(entry: ContentEntry) = {
+    val get = new Get(Bytes.toBytes(entry.getTypeName)).addColumn(META_FAMILY_NAME, schemaCQ)
+    val result = catalogTable.get(get)
+    val spec = Bytes.toString(result.getValue(META_FAMILY_NAME, schemaCQ))
+    SimpleFeatureTypes.createType(entry.getTypeName, spec)
+  }
+
+  override def createTypeNames(): util.List[Name] = {
+    // read types from catalog
+    val scan = new Scan().addColumn(META_FAMILY_NAME, schemaCQ)
+    val scanner = catalogTable.getScanner(scan)
+    scanner.iterator().map { r => Bytes.toString(r.getRow) }.map { n => new NameImpl(n) }.toList
+  }
+}
+
+class HBaseDataStoreFactory extends AbstractDataStoreFactory {
```

Comment:
Hmm... I think we should be implementing DataStoreFactorySpi directly?  (Given that AbstractDataStoreFactory is deprecated.)  That's a later thing though...